docs/spec/crisis/README.md

```diff
@@ -0,0 +1,82 @@
+# WIP 
+
+this spec is WIP and simply an overview of the work to be done in the PR.
+Here is the current thinking as to how the Invariant-Broken-Tx PR will be operate:
+
+Within each module invariant registration functions are defined, for example: 
+
+``` golang
+func RegisterInvariants(ak auth.AccountKeeper, ck types.CrisisKeeper) {
+	ck.RegisterInvariantRoute(ModuleName +"/nonnegative-balance",NonnegativeBalanceInvariant(ak))
+	ck.RegisterInvariantRoute(ModuleName +"/some-other-invariant",SomeOtherInvariant(ak))
+}
+```
+
+Within `cmd/gaia/app/app.go` function `NewGaiaApp` the crisis keeper would be
+initialized with all of the invariants from other modules.  
+
+XXX QUESTION:
+because the invariants are registered in `NewGaiaApp` everything would get
+reregistered in the case of a blockchain reinitialization correct? So these
+routes should not be required to be stored in state. 
+
+The crisis keeper would look something like this: 
+
+``` golang
+type InvarRoute struct {
+	Route     string
+	Invariant sdk.Invariant
+}
+
+type Keeper struct {
+	routes      []InvarRoute
+	distrKeeper DistrKeeper
+}
+
+//expected distribution keeper
+type DistrKeeper interface {
+	MustSpendFeePool(ctx sdk.Context,amt sdk.Coins, addr sdk.AccAddress) error
+}
+```
+
+here the `MustSpendFeePool` is a new function to be created in the distribution
+keeper which which moves coins from `CommunityPool` to an address. It is a "Must"
+function because, if the community pool is empty, the community pool should
+send as much as it has left to the address (and report an error). 
+
+the crisis module will also hold a param in the param store which is the
```

Comment:
I think the incentivization concept is interesting, but I'm not convinced we should do it initially, as opposed to just charging a flat fee & refunding it if an invariant is broken

Reply:
charge a flat fee?? Are you referring to an addtional fee on-top of the gas cost - I think the gas cost should be enough

Reply:
No, I don't think so - the price would be determined by the gas cost * the validator's minimum gas price, but the gas cost we would have to set for that price to be high enough will likely exceed the block gas limit.

Reply:
I'm confused about this last statement, you've described the how the standard gas pricing model work (which is applicable to an invar-broken tx)
1) what does this have to do with the flat fee described earlier in this thread? 
2) "the gas cost we would have to set for that price to be high enough will likely exceed the block gas limit" - why would we set the gas cost? the sender of the Tx determines the cost

Reply:
My point is that the fair amount of gas to charge for this transaction (for a fee of hundreds of USD, say) is likely far higher than the maximum block gas limit, so just setting that gas price won't allow the transaction to be included at all.

Reply:
Well it depends, The idea is to break down all the invariance tests to smaller more specific tests which shouldn't break the gas limit - I think we should just figure out what the limit for account/delegations/etc. size is for this gas limit to be broken. 

Alternatively we could develop some kind exceptions to the gas, for validators who could get slashed if the cry-wolf. 

Alternatively, Alternatively we could remove this from being a transaction - and somehow allow validators to submit some kind of trigger during begin block through config or something (ugly)

Reply:
Running with invariants enabled is something that would place higher resource requirements on the validator so the config might actually make sense. We could even add the ability to enable the invariants on 1 by 1 basis (i.e. a line for each of the invariants). I think that might be cleaner. Are there additional downsides there that I'm not thinking of?

Reply:
This is already the plan, when you submit a borken-invar-tx you must also specify which invariance path is broken.

ALSO another idea, which I now like the best so far (although maybe not super easy to implement): 
Allow for a large invariance to be calculated over multiple txs. We could carry forward an invariance state in inter-block storage between transactions for completion. One would then submit a pot of coins with the transaction to pay the gas price for future transactions if more than 1 transaction is required. Implementing this means we can maintain a consistent ruleset for gas limits - which reduces the possibility of locking up the blockchain (for example a spam transaction in the form of an invariance checks if they were allowed to go over the message gas limit)

Reply:
That is an interesting concept, but it seems unnecessarily complex for a feature that should be as simple as possible. I don't think we want validators to be able to trigger invariants (without paying a fee) either, otherwise it's a way for better-hardware-endowed validators to punish less-well-endowed ones.

I still like the idea of a large flat fee (hundreds of Atoms) - that seems like the easiest spam-prevention mechanism, and it's easy to implement.

Reply:
@cwgoes is the idea that the large flat fee replaces all the gas-fees?

Reply:
Yes, precisely.

Reply:
do we already have the capability to accomplish this? Also notice that in my current implementation the validators are paid for running the calculation, and then the sender is refunded out of the community pool... I'm assuming this would remain the same for the "large fee"

Reply:
We can use `bankKeeper.SubtractCoins` to burn the tokens, then `feeKeeper.AddCollectedFees` to pay them as fees to the validators for running the computation.